ibmdbpy/base.py

```diff
@@ -188,28 +188,45 @@ def __init__(self, dsn, uid='', pwd='', autocommit=True, verbose=False):
 
         if self._con_type == 'jdbc':
 
+            missingCredentialsMsg = ("Missing credentials to connect via JDBC.")
+            ambiguousDefinitionMsg = ("Ambiguous definition of userID or password: " +
+                                      "Cannot be defined in uid and pwd parameters " +
+                                      "and in jdbc_url_string at the same time.")
+
             # remove trailing ":" or ";" or spaces on dsn, we will replace.
             dsn = dsn.rstrip(';: ')
             # find parameters on dsn; if any exist, there will be an equals sign.
             ix = dsn.find("=")
             # if no parameters exist, then this is the complete dsn
             if (ix < 0):
-                # confirm uid and pwd args are present, do NOT allow blank password
-                if (uid == '') | (pwd == ''):
-                    message = ("Missing credentials to connect via JDBC.")
-                    raise IdaDataBaseError(message)
-                # add uid, pwd parameters.
-                dsn = dsn + ':user={};password={};'.format(uid, pwd)
+                # nothing needs to be done, if uid and pwd are missing
+                # (uid and pwd are not needed for local database connections)
+                # otherwise both uid and pwd have to be specified
+                if (uid != '') | (pwd != ''):
+                    # confirm both uid and pwd args are present
+                    if (uid == '') | (pwd == ''):
+                        raise IdaDataBaseError(missingCredentialsMsg)
+                    else:
+                        # add uid, pwd parameters.
+                        dsn = dsn + ':user={};password={};'.format(uid, pwd)
```

Comment:
It took me a while to understand the double if statement, mostly because I find `or` harder to parse than `and`. I find the following easier to read:

```suggestion
                if uid and pwd:
                    dsn = dsn + ':user={};password={};'.format(uid, pwd)
                elif not uid and not pwd:
                   # Neither UID nor PWD have to be specified for local connections.
                   # This assumes that if they're missing, the connection is local.
                    pass
                else:
                     raise IdaDataBaseError(missingCredentialsMsg)                        
```

That is, if there are any parameters in the DSN, either both UID and PWD are required or neither of them.